### PR TITLE
ci: use commitlint to validate messages

### PR DESCRIPTION
**Pull request by marionebl**
This change picks up on the description in this commit: https://github.com/angular/angular/commit/dff6ee3

> Ideally we should switch over to something like https://www.npmjs.com/package/commitplease
as suggested in #9953 but that package currently doesn't support strict scope checking,
which is one of the primarily goal of this PR.

As commitlint is a specialized tool that includes strict scope checking as mentioned in the commit, I thought it might be desirable to replace the tooling code with less complex tool usage.

There is no issue with the current implementation aside from being a non-trivial amount of code. This does not change the current commit validation behaviour.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] ~~Tests for the changes have been added (for bug fixes / features)~~
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[x] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

## Other information

* Picks up on comments in angular/angular#dff6ee3
* Replaces project-internal tooling
* Motivated by commitlint/commitlint#119

### CODE REVIEW

**Comment by marionebl** [comment]
I signed it!

File: scripts/ci/test-js.sh [hunk 1]
```diff
@@ -6,19 +6,11 @@ set -u -e -o pipefail
 readonly thisDir=$(cd $(dirname $0); pwd)
 source ${thisDir}/_travis-fold.sh
 
-
-# Run unit tests for our tools/ directory
-travisFoldStart "test.unit.tools"
```
Comment (marionebl) [inline scripts/ci/test-js.sh]: With the removal of the `validate-commit-message` implementation tand tests this is obsolete, right?

File: tools/gulp-tasks/validate-commit-message.js [hunk 2]
```diff
@@ -9,8 +9,8 @@
 
 // tslint:disable:no-console
 module.exports = (gulp) => () => {
-  const validateCommitMessage = require('../validate-commit-message');
   const shelljs = require('shelljs');
+  const bin = resolveBin('@commitlint/cli', 'commitlint');
```
Comment (gkalpak) [inline tools/gulp-tasks/validate-commit-message.js]: Couldn't this just be `yarn commitlint`?

Comment (marionebl) [reply tools/gulp-tasks/validate-commit-message.js]: Yes, given we'd add a `commitlint` alias to package.json

Comment (gkalpak) [reply tools/gulp-tasks/validate-commit-message.js]: Why would you need an alias?

Comment (marionebl) [reply tools/gulp-tasks/validate-commit-message.js]: Ah i see your point now, are you suggesting we use `commitint` instead of `@commitlint/cli`? I am the author of those packages - `commitlint` mainly exists to maintain backwards compat for now and I suggest usage of `@commitlint/cli` whereever possible.

Comment (gkalpak) [reply tools/gulp-tasks/validate-commit-message.js]: Sorry, I am confused. AFAICT (from a quick) look, `@comitlint/cli` adds a `comittlint` executable in `node_modules/.bin/`. So, wouldn't `yarn comitlint ...` refer to the executable added by `@commitlint/cli`?

How are you suggesting people use `@commitlint/cli`?

**Comment by vicb** [comment]
Why this change ? 
What issue is solved by this PR ? (it should be part of the initial message)

**Comment by marionebl** [comment]
@vicb I updated the PR description, inlining the relevant information from https://github.com/angular/angular/commit/dff6ee3.

File: tools/gulp-tasks/validate-commit-message.js (hunk 2 above)
Comment (marionebl) [reply tools/gulp-tasks/validate-commit-message.js]: Ok sorry, I did not know about `yarn` looking up executables in `node_modules/.bin` for `yarn <command>`. TIL

Comment (marionebl) [reply tools/gulp-tasks/validate-commit-message.js]: Simplified according to your suggestion.

File: scripts/ci/test-js.sh [hunk 3]
```diff
@@ -6,25 +6,11 @@ set -u -e -o pipefail
 readonly thisDir=$(cd $(dirname $0); pwd)
 source ${thisDir}/_travis-fold.sh
 
-
```
Comment (gkalpak) [inline scripts/ci/test-js.sh]: Please, do not remove unrelated whitespace.
(Or at least be consistent about throughout the file :smiley:)
(But better not to remove it in the first place :grin:)

**Review by gkalpak — COMMENTED** [review]
AFAICT (by looking as [@commitlint/config-angular][1]), this change will not keep the same rules as we had before. Could you list all the differences so it is easier to make a decision.
Also, any such differences we decide to enforce should be documented in CONTRIBUTING.md.

[1]: https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js

**Comment by marionebl** [comment]
| Rule | commitlint | validate-commit-message | contributing |
|:-----|:-----------|:---------------------------|:------------:|
| Max. chars in header | [✅](https://github.com/marionebl/angular/blob/62cf2d354c36f7b57b32b14a90cc002863d2ec74/commitlint.config.js#L4) 120 |[✅](https://github.com/angular/angular/blob/master/tools/validate-commit-message/commit-message.json#L2) 120 | [:warning:](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format) 100 |
| Leading capitalized character not allowed in subject | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L9) | ⚠️ | [✅](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#subject) |
| Subject may not be empty* | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L14) | ⚠️* | [⚠️](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#subject) |
| Subject may not end with full stop | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L15) | ⚠️ | [✅](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#subject) |
| Type must be lowercase | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L16) | ⚠️ | ⚠️** |
| Type may not be empty | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L17) | ⚠️* | ⚠️ |
| Allowed types | ⚠️ `'build', 'ci', 'docs', 'feat', 'fix', 'perf', 'refactor', 'revert', 'style', 'test'` ([Source](https://github.com/marionebl/commitlint/blob/ee6ec6eb6bc1b5374bf67a4282cdbee7c9ab3c3d/%40commitlint/config-angular-type-enum/index.js#L1)) | `"build", "ci", "docs", "feat", "fix", "perf", "refactor", "release", "style", "test"` ([Source](https://github.com/angular/angular/blob/0444e13efbebc8fad9373805a6f0287ab2af900d/tools/validate-commit-message/commit-message.json#L3)) + `revert` ([Source](https://github.com/angular/angular/blob/0444e13efbebc8fad9373805a6f0287ab2af900d/tools/validate-commit-message/validate-commit-message.js#L23)) | ⚠️ `"build", "ci", "docs", "feat", "fix", "perf", "refactor", "style", "test"` ([Source](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#type)) + `revert` ([Source](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#revert)) |
| Allowed scopes | ` 'aio', 'animations', 'bazel', 'benchpress', 'common', 'compiler', 'compiler-cli', 'core', 'forms', 'http', 'language-service', 'platform-browser', 'platform-browser-dynamic', 'platform-server', 'platform-webworker', 'platform-webworker-dynamic', 'router', 'service-worker', 'upgrade', 'packaging', 'changelog'` [✅](https://github.com/angular/angular/pull/20521/files#diff-db5313893aaaea45db8e5439f6b81a3cR8) | ` "aio", "animations", "bazel", "benchpress", "common", "compiler", "compiler-cli", "core", "forms", "http", "language-service", "platform-browser", "platform-browser-dynamic", "platform-server", "platform-webworker", "platform-webworker-dynamic", "router", "service-worker", "upgrade", "packaging", "changelog"` [✅](https://github.com/angular/angular/blob/0444e13efbebc8fad9373805a6f0287ab2af900d/tools/validate-commit-message/commit-message.json#L15) | `"aio", "animations", "changelog", "common", "compiler-cli", "compiler", "core", "forms", "http", "language-service", "packaging", "platform-browser-dynamic", "platform-browser", "platform-server", "platform-webworker-dynamic", "platform-webworker", "router", "service-worker", "upgrade",` [:warning:](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#scope) |
| Blank line before body | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L5) | :warning: | [✅](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format) |
| Blank line before footer | [✅](https://github.com/marionebl/commitlint/blob/master/@commitlint/config-angular/index.js#L6) | :warning: | [✅](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format) | |

*May implicitly have been checked by the RegeExp in validate-commit-message
**All allowed types are lowercase, may be implicity defined

**Comment by marionebl** [comment]
@gkalpak I collected the requested information and added a quick summary. Also solved the first Merge Conflicts. Could you have another look?

**Comment by gkalpak** [comment]
Amazing summary 💯 Thx, @marionebl! Taking a look now :smiley:

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -170,7 +170,7 @@ format that includes a **type**, a **scope** and a **subject**:
 
 The **header** is mandatory and the **scope** of the header is optional.
 
-Any line of the commit message cannot be longer 100 characters! This allows the message to be easier
+Any line of the commit message cannot be longer 120 characters! This allows the message to be easier
```
Comment (gkalpak) [inline CONTRIBUTING.md]: Since you are at it: `longer 120` --> `longer than 120` :grin:

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -192,13 +192,14 @@ If the commit reverts a previous commit, it should begin with `revert: `, follow
...
-* **build**: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-* **ci**: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
+* **build**: Changes that affect the build system or external dependencies
+* **ci**: Changes to our CI configuration files and scripts 
 * **docs**: Documentation only changes
 * **feat**: A new feature
 * **fix**: A bug fix
 * **perf**: A code change that improves performance
 * **refactor**: A code change that neither fixes a bug nor adds a feature
+* **release**: Changes related to releases, such as version bumps and changelog additions
```
Comment (gkalpak) [inline CONTRIBUTING.md]: @IgorMinar, is this description accurate?

File: commitlint.config.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+  rules: {'header-max-length': [2, 'always', 120], 'scope-enum': [2, 'always', scopes]}
+};
+
+function getPackages() {
+  return glob.sync('packages/*/package.json')
+      .map(manifest => require(`./${manifest}`).name)
+      .map(name => {
+        const f = name.split('/');
+        return f[f.length - 1];
+      });
```
Comment (gkalpak) [inline commitlint.config.js]: Nit: This last function could be `.map(name => name.split('/').pop())`
<sub>(Sorry, I can't help myself some times 😳)</sub>

**Review by gkalpak — COMMENTED** [review]
Regarding the rules in `validate-commit-message`:

- `Subject may not be empty` was indeed enforced by the RegExp in `validate-commit-message`.
- `Type must be lowercase` was also implicitly enforced by means of white-listd types.
- `Type may not be empty` was also enforced by the RegExp.

Regarding the TODOs:

- `type-case: lower-case` will be explicitly enforced, so I think it is fine to keep it enforced (if it is less configuration) for now.
- `subject-empty` should also be enforced (i.e. the subject should _not_ be empty).
- Not sure what `type-empty` means. To be clear:
  We need to allow omitting the type altogether along with the enclosing parenthesis, but not keeping the parenthesis and omitting the type. I.e. `fix: fix something` is valid, but `fix(): fix something` is not.

##
Other than that, it LGTM (fwiw) 👍

**Comment by marionebl** [comment]
> Regarding the TODOs:

>    type-case: lower-case will be explicitly enforced, so I think it is fine to keep it enforced (if it is less configuration) for now.

That is what the current config is doing. 👍 

>    subject-empty should also be enforced (i.e. the subject should not be empty).
    Not sure what type-empty means. To be clear:
    We need to allow omitting the type altogether along with the enclosing parenthesis, but not keeping the parenthesis and omitting the type. I.e. fix: fix something is valid, but fix(): fix something is not.

`type-empty` affects the first word in `type(scope): subject`. With the current configuration:

* `: some subject` fails
* `type(scope): subject` works
* `type: some subject` works

If I understand you correctly that fits your expectations, right?

**Comment by gkalpak** [comment]
> If I understand you correctly that fits your expectations, right?

Yes, as long as `type(): some subject` also fails.

**Comment by marionebl** [comment]
>> If I understand you correctly that fits your expectations, right?
>
> Yes, as long as type(): some subject also fails.

Yes, it does.

**Comment by marionebl** [comment]
I resolved the conflicts that cropped up in the meantime. Anything holding back landing this PR?

**Comment by marionebl** [comment]
Pretty sure the test failure here is unrelated: https://travis-ci.org/angular/angular/jobs/323050866#L2489

**Review by gkalpak — APPROVED** [review]

**Comment by marionebl** [comment]
The experience of landing this PR has been very unsatisfying. The change have been sitting around in a mergeable state quite some time. 

In that context the wording you chose for the mergeassistant bot borders on passive/agressive from my perspective 

👎

**Comment by marionebl** [comment]
I resolved the merge conflicts that were introduced in the meantime and made sure to add the new `ivy` scope to the commitlint config

**Comment by IgorMinar** [comment]
I've scheduled this PR for our next weeks sprint. @marionebl I'm sorry about the lack of attention from our end, but we have a lot of inflight work and your PR was not aligned with those priorities. Changing our dev infrastructure is not a trivial task and any change requires a through review. I hope we'll get better at this soon. In the meantime please accept my apology.

**Comment by marionebl** [comment]
Closing this due to lack of interest. Won't see the the new merge conflicts / CI issues through. 
Feel free to reuse/reopen if you want to solve yourself.
